### PR TITLE
[DPE-4703] - chore: sync vm + k8s w. nodeport feature

### PR DESCRIPTION
**Pull request by marcoppenheimer**
#### `refactor: sync vm + k8s w. nodeport feature`
- `events/broker.py` and `events/balancer.py` both had started to drift from eachother in term of feature coverage. There were also some minor fixes found dotted around that caused a discrepancy in behavior between the two charms
- These differences in `src/core`, `src/managers`, `src/events` and `tests/unit` have been ported over, along with the (not used on VM) NodePort feature
- In practicality, this has a few downsides:
    - Adding `lightkube` to `pyproject.toml`
    - Adding `expose-external` to the VM config (unavoidable to share `structured_config`)
    - Slightly clunky `if self.charm.substrate == "vm" | "k8s"` dotted about in `events/balancer.py` and `events/broker.py`, for example in observing `kafka_pebble_ready` --> `on_start`
    - Having `health.py` on K8s, which does nothing. Probably the most annoying, and we might have to revisit this file and see if it can be moved to `managers`, and be instantiated cleaner

#### `refactor: general tidy-up of shared components`
- Moved `get_version` to `src/core/workload.py`, as it was now identical on both substrates
- Added `layer` abstractmethod to the core workload, with implementation only on K8s
- Fix the `upgrade_supported` to reflect ZK's removal in 4.0+

### CODE REVIEW

File: tests/unit/test_charm_balancer.py [hunk 1]
```diff
@@ -67,16 +85,17 @@ def test_stop_workload_if_not_leader(patched_restart, charm_configuration):
         config=charm_configuration,
         actions=ACTIONS,
     )
-    state_in = State(leader=False, relations=[])
+    state_in = base_state.replace(leader=False)
```
Comment (zmraul) [inline tests/unit/test_charm_balancer.py:88]: **praise**:  TIL about the `replace` method. This is an excellent way to keep tests similar between k8s and vm

**Review by zmraul — COMMENTED** [review]
Great PR. Had some TILs while reviewing which is always nice :)

File: src/core/workload.py [hunk 2]
```diff
@@ -87,12 +90,12 @@ def jmx_prometheus_javaagent(self):
...
     def jmx_prometheus_config(self):
         """The configuration for the Kafka JMX exporter."""
-        return f"{BROKER.paths['CONF']}/jmx_prometheus.yaml"
+        return f"{self.conf_path}/jmx_prometheus.yaml"
 
     @property
     def jmx_cc_config(self):
         """The configuration for the CruiseControl JMX exporter."""
-        return f"{BALANCER.paths['CONF']}/jmx_cruise_control.yaml"
+        return f"{self.conf_path}/jmx_cruise_control.yaml"
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/core/workload.py:98]: Oh interesting, these were specifically Broker + Balancer, great, thank you!

File: src/events/balancer.py [hunk 3]
```diff
@@ -61,12 +82,19 @@ def __init__(self, charm) -> None:
 
         self.framework.observe(getattr(self.charm.on, "rebalance_action"), self.rebalance)
 
-    def _on_install(self, _) -> None:
+    def _on_install(self, event: InstallEvent) -> None:
         """Handler for `install` event."""
+        if (
+            self.charm.substrate == "k8s"
+            and not self.charm.unit.get_container(CONTAINER).can_connect()
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/events/balancer.py:89]: Good idea, I prefer it.

File: src/events/broker.py [hunk 4]
```diff
@@ -152,15 +211,39 @@ def _on_config_changed(self, event: EventBase) -> None:
...
+                    f"NEW SANs = {expected_sans_ip - current_sans_ip}"
+                )
+            )
+            self.charm.tls.certificates.on.certificate_expiring.emit(
+                certificate=self.charm.state.unit_broker.certificate,
+                expiry=datetime.now().isoformat(),
+            )  # new cert will eventually be dynamically loaded by the broker
+            self.charm.state.unit_broker.update(
+                {"certificate": ""}
+            )  # ensures only single requested new certs, will be replaced on new certificate-available event
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/events/broker.py:243]: It sort-of already is, as we're emitting an event that is handled in the `tls` event-handler. I get your point though that it's a little specific.
How might you improve it do you think?

File: src/events/zookeeper.py [hunk 5]
```diff
@@ -86,23 +89,24 @@ def _on_zookeeper_changed(self, event: RelationChangedEvent) -> None:
...
 
         logger.info(f'Broker {self.model.unit.name.split("/")[1]} disconnected')
         self.charm._set_status(Status.ZK_NOT_RELATED)
 
         # Kafka keeps a meta.properties in every log.dir with a unique ClusterID
         # this ID is provided by ZK, and removing it on relation-broken allows
         # re-joining to another ZK cluster.
-        for storage in self.charm.model.storages[STORAGE]:
-            self.dependent.workload.exec(["rm", f"{storage.location}/meta.properties"])
+        for storage in self.charm.model.storages["data"]:
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/events/zookeeper.py:108]: image

Copied from K8s 👀 

But np, will change both to `STORAGE` :)

[the remark this answers is not in the document]
Comment (zmraul) [reply src/events/zookeeper.py:108]: interesting... :see_no_evil:

File: src/managers/config.py [hunk 6]
```diff
@@ -631,6 +628,7 @@ def set_environment(self) -> None:
         updated_env_list = [
             self.kafka_opts,
             self.kafka_jmx_opts,
+            self.cc_jmx_opts,
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/managers/config.py:631]: I suppose this gets 'hidden' as we likely `broker.set_environment` THEN `balancer.set_environment`, which map_env overwrites that key, right?
Will remove.

[the remark this answers is not in the document]
Comment (zmraul) [reply src/managers/config.py:631]: Yes, we found a situation where broker is the last to invoke `set_environmnet` and we ended up with a wrong value.

File: src/managers/config.py [hunk 7]
```diff
@@ -770,7 +768,7 @@ def cruise_control_properties(self) -> list[str]:
                 "zookeeper.security.enabled=true",
                 f'sasl.jaas.config=org.apache.kafka.common.security.scram.ScramLoginModule required username="{self.state.balancer.broker_username}" password="{self.state.balancer.broker_password}";',
                 "sasl.mechanism=SCRAM-SHA-512",
-                f"security.protocol={self.security_protocol}",
+                f"security.protocol={self.state.default_auth}",
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply src/managers/config.py:771]: Good catch!
And probably on the `mechanism` part, have added.

File: src/core/workload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Batalex) [reply src/core/workload.py:98]: Sorry about that, I should have added a comment to clarify this behaviour

**Review by Batalex — APPROVED** [review]
Great job! Unifying the sub operator looked like a challenge. I don't have much to add after reviewing the k8s PR and Raúl's comments

File: tests/unit/test_balancer.py [hunk 8]
```diff
@@ -16,8 +16,6 @@
 
 logger = logging.getLogger(__name__)
 
-pytestmark = pytest.mark.balancer
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply tests/unit/test_balancer.py:19]: I don't recall removing it explicitly, could you remind what it does? Is it so we can run:
```
tox -e unit -- -m "balancer"
```
?

File: tests/unit/test_charm_balancer.py [hunk 9]
```diff
@@ -160,9 +180,10 @@ def test_ready_to_start_no_zk_data(charm_configuration):
     assert state_out.unit_status == Status.ZK_NO_DATA.value.status
 
 
-def test_ready_to_start_no_broker_data(charm_configuration, zk_data):
+def test_ready_to_start_no_broker_data(charm_configuration, base_state: State, zk_data):
     # Given
     charm_configuration["options"]["roles"]["default"] = "balancer,broker"
+    charm_configuration["options"]["expose-external"]["default"] = "none"
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply tests/unit/test_charm_balancer.py:186]: Very.

File: tests/unit/test_config.py [hunk 10]
```diff
@@ -29,23 +30,15 @@
 )
 from managers.config import ConfigManager
 
-pytestmark = [pytest.mark.broker, pytest.mark.balancer]
+pytestmark = pytest.mark.broker
```
[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply tests/unit/test_config.py:33]: Same comment as https://github.com/canonical/kafka-operator/pull/226#discussion_r1717107736

File: tests/unit/test_balancer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Batalex) [reply tests/unit/test_balancer.py:19]: Yes, that's right

[the remark this answers is not in the document]
Comment (marcoppenheimer) [reply tests/unit/test_balancer.py:19]: I'll add them back.

File: src/events/broker.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zmraul) [reply src/events/broker.py:237]: I would expect that setting `{"certificate": ""}` should be done within the handler instead of here for example.

**Review by zmraul — APPROVED** [review]

**Pull request merged**